go-controller/pkg/ovn/ovn.go

```diff
@@ -1047,148 +1041,66 @@ func (oc *Controller) syncNodeGateway(node *kapi.Node, hostSubnets []*net.IPNet)
+				if err := oc.deleteNodeEvent(node); err != nil {
+					oc.retryNodes.unSkipRetryObj(node.Name)
+					klog.Errorf("Failed to delete node %s, error: %v",
+						node.Name, err)
+					return
 				}
-				return
+				oc.retryNodes.removeDeleteFromRetryObj(node.Name)
 			}
-
-			klog.V(5).Infof("Added event for Node %q", node.Name)
-			hostSubnets, err := oc.addNode(node)
-			if err != nil {
-				klog.Errorf("NodeAdd: error creating subnet for node %s: %v", node.Name, err)
-				addNodeFailed.Store(node.Name, true)
-				nodeClusterRouterPortFailed.Store(node.Name, true)
-				mgmtPortFailed.Store(node.Name, true)
-				gatewaysFailed.Store(node.Name, true)
+			start := time.Now()
+			if err := oc.addUpdateNodeEvent(node, nil); err != nil {
+				klog.Errorf("Failed to create node %s, error: %v",
+					node.Name, err)
+				oc.retryNodes.unSkipRetryObj(node.Name)
 				return
 			}
-
-			if err = oc.syncNodeClusterRouterPort(node, hostSubnets); err != nil {
-				if !util.IsAnnotationNotSetError(err) {
-					klog.Warningf(err.Error())
-				}
-				nodeClusterRouterPortFailed.Store(node.Name, true)
-			}
-
-			err = oc.syncNodeManagementPort(node, hostSubnets)
-			if err != nil {
-				if !util.IsAnnotationNotSetError(err) {
-					klog.Warningf("Error creating management port for node %s: %v", node.Name, err)
-				}
-				mgmtPortFailed.Store(node.Name, true)
-			}
-
-			if err := oc.syncNodeGateway(node, hostSubnets); err != nil {
-				if !util.IsAnnotationNotSetError(err) {
-					klog.Warningf(err.Error())
-				}
-				gatewaysFailed.Store(node.Name, true)
-			}
-
-			// ensure pods that already exist on this node have their logical ports created
-			options := metav1.ListOptions{FieldSelector: fields.OneTermEqualSelector("spec.nodeName", node.Name).String()}
-			pods, err := oc.client.CoreV1().Pods(metav1.NamespaceAll).List(context.TODO(), options)
-			if err != nil {
-				klog.Errorf("Unable to list existing pods on node: %s, existing pods on this node may not function")
-			} else {
-				oc.addRetryPods(pods.Items)
-				oc.requestRetryPods()
-			}
-
+			klog.Infof("Created Node: %s took: %v", node.Name, time.Since(start))
+			oc.retryNodes.checkAndDeleteRetryObj(node.Name, true)
 		},
 		UpdateFunc: func(old, new interface{}) {
 			oldNode := old.(*kapi.Node)
-			node := new.(*kapi.Node)
-
-			shouldUpdate, err := shouldUpdate(node, oldNode)
-			if err != nil {
-				klog.Errorf(err.Error())
-			}
-			if !shouldUpdate {
-				// the hostsubnet is not assigned by ovn-kubernetes
-				return
-			}
-
-			var hostSubnets []*net.IPNet
-			_, failed := addNodeFailed.Load(node.Name)
-			if failed {
-				hostSubnets, err = oc.addNode(node)
-				if err != nil {
-					klog.Errorf("NodeUpdate: error creating subnet for node %s: %v", node.Name, err)
+			newNode := new.(*kapi.Node)
+			oc.retryNodes.checkAndSkipRetryObj(newNode.Name)
+			if retryEntry := oc.retryNodes.getObjRetryEntry(newNode.Name); retryEntry != nil && retryEntry.oldObj != nil {
```

Comment:
In the existing code for network policies, in this part we actually consider and process the old object:
https://github.com/ovn-org/ovn-kubernetes/pull/2870/files#diff-796adb9d8a9b5b5636dc81e7350a00ef143e6e5bcb52ec30dc350c866cd9208cR677-R686

Here I see that you are doing everything around the new node. Either one of the two implementations is wrong, I suppose :)

Reply:
I was more of less based off the pods handling since it does have update path unlike netpol

Reply:
nodes is more like pods with update hence I was more aligned with pod update https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/ovn.go#L588

Reply:
I think Ricky means that a few lines down:
```
if err := oc.deleteNodeEvent(newNode); err != nil {
```

^ that should be deleting the old node object, not the new

Reply:
understood but this isn't what pod update do 
https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/ovn.go#L607
its a node that was in cache for delete retry hence we redelete it